src/dune_engine/action_builder.ml

```diff
@@ -504,9 +492,121 @@ struct
+  | Label _ -> true
+  | Or_exn b -> can_eval_statically b
+  | Fail _ -> true
+  | If_file_exists (_, _, _) -> false
+  | Memo _ -> false
+  | Catch (t, _) -> can_eval_statically t
+  | Memo_build _ -> false
+  | Dyn_memo_build _ -> false
+  | Build _ ->
+    (* TODO jeremiedimino: This should be [can_eval_statically b], however it
+       breaks the [Expander.set_artifacts_dynamic] trick that it used to break a
+       cycle. The cycle is as follow:
+
+       - [(rule (deps %{cmo:x}) ..)] requires expanding %{cmo:x}
+
+       - expanding %{cmo:x} requires computing the artifacts DB
+
+       - computing the artifacts DB requires computing the module<->library
+       assignment
+
+       - computhig the above requires knowing the set of source files in a given
+       directory
+
+       - computing the above works by looking at the source tree and adding all
+       targets of user rules
+
+       - computing targets of user rules is done by effectively generating the
+       rules for the user rules, which means interpreting the [(deps
+       %{cmo:...})] thing
+
+       In theory, we should be able to compute the list
+
+       looks odd, if we find a way to break it we should be able to change this. *)
+    false
+
+let static_eval =
+  let rec loop : type a. a t -> unit t -> a * unit t =
+   fun t acc ->
+    match t with
+    | Pure x -> (x, acc)
+    | Map (f, a) ->
+      let x, acc = loop a acc in
+      (f x, acc)
+    | Both (a, b) ->
+      let a, acc = loop a acc in
+      let b, acc = loop b acc in
+      ((a, b), acc)
+    | Seq (a, b) ->
+      let (), acc = loop a acc in
+      let b, acc = loop b acc in
+      (b, acc)
+    | Map2 (f, a, b) ->
+      let a, acc = loop a acc in
+      let b, acc = loop b acc in
+      (f a b, acc)
+    | All xs -> loop_many [] xs acc
+    | Paths_for_rule _ -> assert false
+    | Paths_glob _ -> assert false
+    | Deps _ -> ((), acc >>> t)
+    | Dyn_paths b ->
+      let (x, ps), acc = loop b acc in
+      (x, Deps (Dep.Set.of_files_set ps) >>> acc)
+    | Dyn_deps b ->
+      let (x, deps), acc = loop b acc in
+      (x, Deps deps >>> acc)
+    | Contents _ -> assert false
+    | Lines_of _ -> assert false
+    | Label _ -> ((), acc >>> t)
+    | Or_exn b ->
+      let res, acc = loop b acc in
+      (Result.ok_exn res, acc)
+    | Fail { fail } -> fail ()
+    | If_file_exists (_, _, _) -> assert false
+    | Memo _ -> assert false
+    | Catch (t, v) -> ( try loop t acc with _ -> (v, return ()) )
+    | Memo_build _ -> assert false
+    | Dyn_memo_build _ -> assert false
+    | Build b ->
+      let b, acc = loop b acc in
+      loop b acc
+  and loop_many : type a. a list -> a t list -> unit t -> a list * unit t =
+   fun acc_res l acc ->
+    match l with
+    | [] -> (List.rev acc_res, acc)
+    | t :: l ->
+      let x, acc = loop t acc in
+      loop_many (x :: acc_res) l acc
+  in
+  fun t ->
+    if can_eval_statically t then
+      Some (loop t (return ()))
```

Comment:
This looks odd. I suppose you could get rid of `can_eval_statically` because `loop` essentially duplicates its logic by throwing an exception where `can_eval_statically` returns false. Instead, you could return `None` when catching an exception.

(Alternatively, `loop` could return an `option` but that's probably annoying to code and might be slower?)

Reply:
> (Alternatively, loop could return an option and avoid throwing exceptions but the resulting code might end up being pretty verbose and might also be slower?)

That was my thinking. I feel like we should return `(unit -> 'a) option` rather than `'a option`, so that in the case the action builder cannot be evaluated statically we are sure to execute nothing. But I expect that we will get rid of this altogether eventually, once we have pushed the monad enough. So I don't mind either way, happy to rewrite this in one function.

Reply:
Switching to `(unit -> 'a) option` seems fine. If you prefer to keep the current code for now, with the aim to eventually get rid of duplication, then it seems worth adding a comment about this.

Reply:
Just to clarify, I expect that we will get rid of the `static_eval` entirely in the future. Agreed that we should add a comment about it. I added one.